syncd/PortLinkEventDamper.cpp

```diff
@@ -0,0 +1,416 @@
+                "Expected current accumulated penalty %u to be <= link reuse threshold "
+                "%u "
+                "for port vid: %s, need another %d usecs.",
+                m_accumulatedPenalty, m_reuseThreshold,
+                sai_serialize_object_id(m_vid).c_str(), tv_usec);
+
+        resetTimer(tv_usec);
+    }
+}
+
+void PortLinkEventDamper::handlePortStateChange(
+        _In_ sai_port_oper_status_t newPortState)
+{
+    SWSS_LOG_ENTER();
+
+    m_actualState = newPortState;
+
+    if (newPortState == SAI_PORT_OPER_STATUS_UP)
+    {
+        ++m_stats.pre_damping_up_events;
+    }
+    else
+    {
+        ++m_stats.pre_damping_down_events;
+    }
+
+    // If link event damping config is not enabled, advertise the port event.
+    if (m_dampingConfigEnabled == false)
+    {
+        advertisePortState(newPortState);
+
+        return;
+    }
+
+    if (m_dampingActive == false)
+    {
+        advertisePortState(newPortState);
+
+        // Penalty is updated only for DOWN events.
+        if (newPortState == SAI_PORT_OPER_STATUS_DOWN)
+        {
+            updatePenalty();
+
+            if (m_accumulatedPenalty >= m_suppressThreshold)
+            {
+                // Start the damping on the port.
+                resetTimer(sairedis::Utils::timeToReachTargetValueUsingHalfLife(
+                           m_decayHalfLifeUsec, m_accumulatedPenalty, m_reuseThreshold));
+                m_dampingActive = true;
+
+                SWSS_LOG_INFO(
+                        "Damping starts at time %llu on port vid: %s with accumulated "
+                        "penalty: %u, reuse_threshold: %u",
+                        m_lastPenaltyUpdateTimestamp,
+                        sai_serialize_object_id(m_vid).c_str(), m_accumulatedPenalty,
+                        m_reuseThreshold);
+            }
+        }
+    }
+    else // Damping is active on port.
+    {
+        // UP event on a damped link is no-op.
+        if (newPortState == SAI_PORT_OPER_STATUS_DOWN)
+        {
+            // Cancel the timer on the port.
+            cancelTimer();
+            updatePenalty();
+
+            // Restart the timer.
+            resetTimer(sairedis::Utils::timeToReachTargetValueUsingHalfLife(
+                       m_decayHalfLifeUsec, m_accumulatedPenalty, m_reuseThreshold));
+        }
+    }
+}
+
+void PortLinkEventDamper::advertisePortState(
+        _In_ sai_port_oper_status_t portState)
+{
+    SWSS_LOG_ENTER();
+
+    if (portState != m_advertisedState)
+    {
+        sai_port_oper_status_notification_t data = {.port_id = m_rid,
+                                                    .port_state = portState};
+
+        SWSS_LOG_NOTICE("Advertising [port rid: %s, port oper status: %s]",
+                sai_serialize_object_id(data.port_id).c_str(),
+                sai_serialize_port_oper_status(data.port_state).c_str());
+
+        m_notificationHandler->onPortStateChangePostLinkEventDamping(/*count=*/1, &data);
```

Comment:
this is not right, you are manually generating port notification, this breaks SAI design, @lguohan  we need to discuss this whether this is allowed

Reply:
This is not manual generation of port event notification exactly rather delayed generation of port events as it is intended with link event damping feature - for the link event damping, the actual port events  will be trapped here  (https://github.com/sonic-net/sonic-sairedis/blob/master/syncd/NotificationHandler.cpp#L164-L172) (NotificationHandler changes will be added in subsequent PRs to do that) and sent to link event damping logic running in syncd main thread where events will be processed as per the damping algo and config. If port events are not needed to be damped or damping duration have elapsed and damped port event needs to be advertised to OA (and other applications), corresponding port event notifications will be generated as the case here.

Since we are doing review of a portion of a code at a time, full logic is not self explanatory here.

Targeted final onPortStateChange() will look like following where port events are trapped and sends for processing. Here m_portStateChangeHandler is of type PortStateChangeHandler (added in https://github.com/sonic-net/sonic-sairedis/pull/1310):
===
void NotificationHandler::onPortStateChange(
        _In_ uint32_t count,
        _In_ const sai_port_oper_status_notification_t *data)
{
    SWSS_LOG_ENTER();

    if (m_portStateChangeHandler == nullptr) {
        auto s = sai_serialize_port_oper_status_ntf(count, data);
        enqueueNotification(SAI_SWITCH_NOTIFICATION_NAME_PORT_STATE_CHANGE, s);
    } else {
        // Send notification to port state change handler for further processing.
        m_portStateChangeHandler->HandlePortStateChangeNotification(count, data);
    }
}

=======

cc: @mikeberesford.

Reply:
it seems like manual generate, since this code is creating notification data and manually calling onPortStateChangePostLinkEventDamping, which all notifications should come from vendor SAI and not be generated on demand

Reply:
That is the intention of the feature. SAI generated link event is trapped by link event damping feature and it will be sent to application as per the damping logic. For example:

1) if damping logic says SAI generated event notification needs to damped, that notification will be suppressed in the syncd and will not be sent to application.
2) if damping logic says no damping needed, then the notification needs to be sent to application using onPortStateChangePostLinkEventDamping(). Since original SAI notification is trapped and processed in the damper logic, we need to manually create the corresponding notification data and send it to application.

Since the original notification from SAI is trapped and processed by damper, we will need to create notification data when notification needs to be sent.

Reply:
Not fully understand this logic, can you make some kind of flow diagram ? So e I'm not convinced still why we generate manually this notification